src/main/java/io/github/opencubicchunks/cubicchunks/cubicgen/customcubic/CustomGeneratorSettings.java

```diff
@@ -185,17 +186,74 @@ public BiomeBlockReplacerConfig createBiomeBlockReplacerConfig() {
         return replacerConfig;
     }
 
-    public String toJson(boolean minimize) {
-        Gson gson = gson(minimize);
+    public String toJson() {
+        Gson gson = gson();
         return gson.toJson(this);
     }
+    
+    public static boolean isOutdated(String settingsJsonString) {
+        JsonObject root = CustomGeneratorSettingsFixer.stringToJson(settingsJsonString);
+        return !root.has("version") || root.get("version").getAsInt() != CustomGeneratorSettingsFixer.VERSION;
+    }
 
-    public static CustomGeneratorSettings fromJson(String json) {
-        if (json.isEmpty()) {
+    public static CustomGeneratorSettings fromJson(String jsonString) {
+        if (jsonString.isEmpty()) {
             return defaults();
         }
-        Gson gson = gson(true); // minimize option shouldn't matter when deserializing
-        return gson.fromJson(json, CustomGeneratorSettings.class);
+        if (isOutdated(jsonString)) {
+            jsonString = CustomGeneratorSettingsFixer.fixGeneratorOptions(CustomGeneratorSettingsFixer.stringToJson(jsonString), false);
+        }
+        Gson gson = gson();
+        return gson.fromJson(jsonString, CustomGeneratorSettings.class);
+    }
+    
+    public static CustomGeneratorSettings load(World world) {
+        File externalGeneratorPresetFile = new File(world.getSaveHandler().getWorldDirectory(),
+                "/data/" + CustomCubicMod.MODID + "/custom_generator_settings.json");
+        String jsonString = null;
+        CustomGeneratorSettings settings = null;
+        if (externalGeneratorPresetFile.exists()) {
+            try (FileReader reader = new FileReader(externalGeneratorPresetFile)){
+                CharBuffer sb = CharBuffer.allocate((int) externalGeneratorPresetFile.length() * 2);
```

Comment:
What... why twice the length?

Reply:
`CharBuffer.allocate(int)` consume char buffer capacity in chars, `File.length()` return size in bytes.

Reply:
Well, I'm derping a little. I should divide by 2.

Reply:
Actually no. Files and java chars use completely different character encoding. Files are UTF-8 and java uses UTF-16. For the ASCII subset, 1 byte in file corresponds to 1 java character. There can be multiple bytes in file that also correspond to a single java char. But not the other way around.